Code/Engine/Core/Scripting/Implementation/ScriptComponent.cpp

```diff
@@ -0,0 +1,302 @@
+{
+  ezHashedString hs;
+  hs.Assign(szKey);
+
+  auto it = m_Parameters.Find(hs);
+  if (it != ezInvalidIndex && m_Parameters.GetValue(it) == value)
+    return;
+
+  m_Parameters[hs] = value;
+}
+
+void ezScriptComponent::RemoveParameter(const char* szKey)
+{
+  if (m_Parameters.RemoveAndCopy(ezTempHashedString(szKey)))
+  {
+  }
+}
+
+bool ezScriptComponent::GetParameter(const char* szKey, ezVariant& out_value) const
+{
+  ezUInt32 it = m_Parameters.Find(szKey);
+
+  if (it == ezInvalidIndex)
+    return false;
+
+  out_value = m_Parameters.GetValue(it);
+  return true;
+}
+
+void ezScriptComponent::InstantiateScript(bool bActivate)
+{
+  ClearInstance(true);
+
+  ezResourceLock<ezScriptClassResource> pScript(m_hScriptClass, ezResourceAcquireMode::BlockTillLoaded_NeverFail);
+  if (pScript.GetAcquireResult() != ezResourceAcquireResult::Final)
+  {
+    ezLog::Error("Failed to load script '{}'", GetScriptClassFile());
+    return;
+  }
+
+  auto pScriptType = pScript->GetType();
+  if (pScriptType == nullptr || pScriptType->IsDerivedFrom(ezGetStaticRTTI<ezComponent>()) == false)
+  {
+    ezLog::Error("Script type '{}' is not a component", pScriptType != nullptr ? pScriptType->GetTypeName() : "NULL");
+    return;
+  }
+
+  m_pScriptType = pScriptType;
+  m_pMessageDispatchType = pScriptType;
+
+  m_pInstance = pScript->Instantiate(*this, GetWorld());
+  if (m_pInstance != nullptr)
+  {
+    m_pInstance->ApplyParameters(m_Parameters);
+  }
+
+  UpdateScheduling();
+
+  CallScriptFunction(ezComponent_ScriptBaseClassFunctions::Initialize);
+  if (bActivate)
+  {
+    CallScriptFunction(ezComponent_ScriptBaseClassFunctions::OnActivated);
+  }
+}
+
+void ezScriptComponent::ClearInstance(bool bDeactivate)
+{
+  if (bDeactivate)
+  {
+    CallScriptFunction(ezComponent_ScriptBaseClassFunctions::OnDeactivated);
+  }
+  CallScriptFunction(ezComponent_ScriptBaseClassFunctions::Deinitialize);
+
+  auto pModule = GetWorld()->GetOrCreateModule<ezScriptWorldModule>();
+  if (auto pUpdateFunction = GetScriptFunction(ezComponent_ScriptBaseClassFunctions::Update))
+  {
+    pModule->RemoveUpdateFunctionToSchedule(pUpdateFunction, m_pInstance.Borrow());
+  }
+
+  pModule->RemoveScriptReloadFunction(m_hScriptClass, this);
+
+  m_pInstance = nullptr;
+  m_pScriptType = nullptr;
+
+  m_pMessageDispatchType = GetDynamicRTTI();
+}
+
+void ezScriptComponent::UpdateScheduling()
+{
+  auto pModule = GetWorld()->GetOrCreateModule<ezScriptWorldModule>();
+  if (auto pUpdateFunction = GetScriptFunction(ezComponent_ScriptBaseClassFunctions::Update))
+  {
+    pModule->AddUpdateFunctionToSchedule(pUpdateFunction, m_pInstance.Borrow(), m_UpdateInterval);
+  }
+
+  pModule->AddScriptReloadFunction(m_hScriptClass,
+    [this]() {
+      InstantiateScript(true);
+    });
+}
+
+const ezAbstractFunctionProperty* ezScriptComponent::GetScriptFunction(ezUInt32 uiFunctionIndex)
+{
+  if (m_pScriptType != nullptr && m_pInstance != nullptr)
```

Comment:
is the instance check really needed ?

Reply:
Yes, otherwise you can't call the returned function. You could add the same check in all the places where you use the returned function but it was simpler to do it this way.